src/OpenSearch/Client.php

```diff
@@ -1977,34 +1868,57 @@ public function extractArgument(array &$params, string $arg)
     }
 
     /**
-     * Sends a raw request to the cluster
-     * @return callable|array
-     * @throws NoNodesAvailableException
+     * Check if the client is running in async mode.
      */
-    public function request(string $method, string $uri, array $attributes = [])
+    public function isAsync(): bool
+    {
+        return $this->isAsync;
+    }
+
+    /**
+     * Set the client to run in async mode.
+     */
+    public function setAsync(bool $isAsync): static
+    {
+        $this->isAsync = $isAsync;
+        return $this;
+    }
+
+    /**
+     * Sends a raw request to the cluster.
+     *
+     * @throws \Psr\Http\Client\ClientExceptionInterface|\Exception
+     */
+    public function sendRawRequest(string $method, string $uri, array $params = [], ?string $body = null): Promise|ResponseInterface
     {
         $params = $attributes['params'] ?? [];
         $body = $attributes['body'] ?? null;
-        $options = $attributes['options'] ?? [];
 
-        $promise = $this->transport->performRequest($method, $uri, $params, $body, $options);
+        $request = $this->transport->createRequest($method, $uri, $params, $body);
 
-        return $this->transport->resultOrFuture($promise, $options);
+        if ($this->isAsync()) {
```

Comment:
instead of an instance-level flag (which separates the configuration from the usage, forcing all caller code to actually support both return types as they don't know how the client was configured), maybe this should be a argument of that method, allowing to use a conditional return type to be more precise (or even separate methods)

Reply:
Looking at other clients in other languages, they often provide a *separate* async client (and corresponding async namespaces) for the async case, allowing to have proper return types for both cases instead of a union type.